### ISSUE TITLE
Training larger model is (much) faster than smaller ones 

### ISSUE THREAD

**Issue by jli05** [google/jax#4472]
Below is a toy code. If we use `model_sizes = (500, 20, 20, 100)` instead of `model_sizes = (500, 50, 50, 100)` that'd increase training time by 0.1s just for illustration. For a real application the time difference would be much more significant. Any idea why?

(I put `jax.jit` around the training `one_step()` function.)

``` python3

import jax
import jax.numpy as jnp
import numpy as np

relu = lambda x : jnp.maximum(x, 0)

def factory(key, model_sizes):
    key1, key2, key3 = jax.random.split(key, 3)
    h1, h2, h3, h4 = model_sizes
    m1 = jax.random.normal(key1, (h1, h2))
    m2 = jax.random.normal(key2, (h2, h3))
    m3 = jax.random.normal(key3, (h3, h4))
    params = (m1, m2, m3)

    def model(params, x):
        m1, m2, m3 = params
        y1 = relu(x @ m1)
        y2 = relu(y1 @ m2)
        y3 = relu(y2 @ m3)
        return y3

    def loss(params, x, y):
        y_pred = model(params, x)
        return .5 * jnp.mean((y - y_pred) ** 2)

    return params, model, loss

def train(params, loss, x, y, batch_size=100,
          learning_rate=.05, steps=200):
    grad_fn = jax.grad(loss)

    from jax.experimental.optimizers import sgd
    opt = sgd(learning_rate)
    opt_state = opt.init_fn(params)

    @jax.jit
    def one_step(j, opt_state, x, y):
        grads = grad_fn(opt.params_fn(opt_state), x, y)
        return opt.update_fn(j, grads, opt_state)

    from random import sample
    for j in range(steps):
        batch_index = sample(range(x.shape[0]), batch_size)
        opt_state = one_step(j, opt_state, x[batch_index], y[batch_index])

    return opt.params_fn(opt_state)

if __name__ == '__main__':
    model_sizes = (500, 50, 50, 100)
    params, _, loss = factory(jax.random.PRNGKey(87349), model_sizes)
    x, y = np.random.rand(2000, 500), np.random.rand(2000, 100)
    train(params, loss, x, y)
```

**Comment by jli05** [comment]
Found related Issues #427 #193

I tried rewrite the training loop into `jax.lax.fori_loop()`. The runtime could be much longer and the result accuracy seems worse(-ish). If we write `jax.jit(jax.lax.fori_loop())` an error with long data dump would be thrown.

Thought it would be great if a developer could get acceptable baseline performance "out of box".

**Comment by jli05** [comment]
All the tests are on CPU.

**Comment by skye** [comment]
Can you share how you're doing your timings? I don't see any timing code in your script. Are you using https://jax.readthedocs.io/en/latest/profiling.html ?

**Comment by jli05** [comment]
I just used bash `time` utility.

```bash
time python3 demo_of_issue.py
```

**Comment by jli05** [comment]
Regards the profiling, I added the following code but don't see the "Capture" button at top-left of TensorBoard.

```python3
import time
import jax.profiler
server = jax.profiler.start_server(9999)
time.sleep(5)
```

Screenshot of TensorBoard whilst running the Python application:

<img width="880" alt="Screenshot 2020-10-08 at 13 56 18" src="https://user-images.githubusercontent.com/3519667/95461636-5e6a2800-096e-11eb-9d92-3f543e306871.png">

**Comment by jli05** [comment]
When written in TF 2.3, training smaller model is faster (normal behaviour). No Eager Execution enabled.

JAX 0.2 seems to exhibit the converse of normal behaviour.
